### ISSUE TITLE
Need of a way to pre-fill DB with data at genesis or / and way to add data to the running DB `for testing`

### ISSUE THREAD

**Issue by phklive** [0xPolygonMiden/miden-node#127]
The Notes and Nullifiers tables are empty for now and the only way to be able to fill them up is through the `apply_block` method. 

https://github.com/0xPolygonMiden/miden-node/blob/59984a5d4969053c55edf148c978a30b0ada23e9/store/src/server/api.rs#L100-L130

For testing purposes we could have a way to pre-fill the DB with data at genesis or an easy way to add data while the DB runs.

Linked to: #97 and #126

**Comment by bobbinth** [comment]
I'd expand this issue a bit to make it more broadly about generating genesis files. Currently, all aspects of [generating genesis files](https://github.com/0xPolygonMiden/miden-node/blob/main/node/src/commands.rs#L61) are hard-coded in Rust. We should make it more flexible. I can think of the following improvements:

### Human-readable inputs
It would be great if we could supply inputs for `make-genesis` command via some human-readable file. This could be a JSON file of the following form:
```JSON
{
    "version": 0,
    "timestamp": 13456,
    "accounts": [
        {
            "type": "basic_wallet",
            "id": "0x123456",
            "seed": "0x1245",
            "auth": {
                "type": "RpoFalcon512",
                "seed": "0x12456"
            },
            "storage_mode": "local"
        },
        {
            "type": "fungible_faucet",
            "id": "0x123456",
            "seed": "0x12345",
            "token_symbol": "POL",
            "decimals": 9,
            "max_supply": 1000000000,
            "auth": {
                "type": "RpoFalcon512",
                "seed": "0x12456"
            },
            "storage_mode": "onchain"
        }
    ],
    "notes": [],
    "nullifiers": []
}
```
(we could also adapt other formats if it simplifies parsing)

The procedure would then parse the file, create the specified accounts, notes, and nullifiers (using procedures exposed in `miden-lib`), and then output the genesis file as it does now.

### Account exports
Right now, the genesis generation procedure exports just public-private key pairs for keys associated with with the accounts in the genesis file. Instead, I think we should output files which can then be easily imported into the client. Specifically, we can output one file per account - maybe with `.mac` extension.

We need to settle on the format for these files. At the very least account file should contain:
- Serialized `Account` struct.
- Optional `account_seed` that we get at account creation time.
- Account authentication info (i.e., schema + key info). Here, we might be fine with serializing schema + the seed from which private/public keys were derived.

But @igamigo can provide more input on whether anything else is needed for importing the account into the client.

**Comment by phklive** [comment]
@bobbinth, for now the `make_genesis` [function](https://github.com/0xPolygonMiden/miden-node/blob/3f64cf3413d358033f0b52acc1a8fd0cf5ed14d1/node/src/commands.rs#L61) creates 2 accounts by default: 
- Basic wallet 
- Basic fungible faucet

I have 4 questions with the addition of a genesis file (that would look like genesis.json or other file extension) :
- Should we make the genesis file TOML to keep the use of toml files as in miden.toml or more generally in rust? 
```toml
version = 0
timestamp = 13456
accounts = [
    {type = "basic_wallet", id = "0x123456", seed = "0x1245", storage_mode = "local", auth = {type = "RpoFalcon512", seed = "0x12456"}},
    {type = "fungible_faucet", id = "0x123456", seed = "0x12345", token_symbol = "POL", decimals = 9, max_supply = 1000000000, storage_mode = "onchain", auth = {type = "RpoFalcon512", seed = "0x12456"}}
]
notes = []
nullifiers = []
```
- Should we make the genesis file optional? 
- If the genesis file is optional, then should we keep the 2 accounts initialised at genesis or should we start from a blank state? 
- Why do we have the word `basic` in front of wallet and faucet? Does it mean that there is a `not_basic` faucet or wallet that have a different behaviour?

**Comment by bobbinth** [comment]
> Should we make the genesis file TOML to keep the use of toml files as in miden.toml or more generally in rust?

I'm fine with using TOML (especially since we can put comments in there) - but we should probably use a formatted version so that we don't get super long lines.

> Should we make the genesis file optional?

I'm inclined not to make it optional, a least for now.

> Why do we have the word `basic` in front of wallet and faucet? Does it mean that there is a `not_basic` faucet or wallet that have a different behaviour?

These correspond to the account construction functions we have in [miden-lib](https://github.com/0xPolygonMiden/miden-base/tree/main/miden-lib/src/accounts). Right now, we don't have anything but "basic" wallets an faucets but this will probably change in the future.

**Comment by phklive** [comment]
@bobbinth, Could you explain to me why we have our own implementations of `Serializable` and `Deserializable` for `GenesisState` and `Account`? 

- https://github.com/0xPolygonMiden/miden-node/blob/3f64cf3413d358033f0b52acc1a8fd0cf5ed14d1/store/src/genesis.rs#L59-L89

- https://github.com/0xPolygonMiden/miden-base/blob/5390a2cc099727dd6cd8ffc3ae5065a27ab3d4a4/objects/src/accounts/mod.rs#L169-L194

Account also implements Serde Serialize and Deserialize which I believe that I will need to be able to parse the genesis.toml file and turn it into a valid GenesisState type.

**Comment by igamigo** [comment]
Nullifiers being Digests should already be serializable although it should be better to implement it explicitly. Probably notes should also have a serialization implementation if we want to make them human-readable. I wonder if we want to also save notes as binary files and reference them from the state JSON. Note that the client can already import/export note binary files (for exchanging between clients) and so it might make sense to consolidate ways to exchange this type of data across projects.

Regarding accounts, not sure if saving account seed is enough to easily derive the key pair info but if it is, I agree with Bobbin's proposal regarding saving a specific file per account.

**Comment by phklive** [comment]
> Nullifiers being Digests should already be serializable although it should be better to implement it explicitly. Probably notes should also have a serialization implementation if we want to make them human-readable. I wonder if we want to also save notes as binary files and reference them from the state JSON. Note that the client can already import/export note binary files (for exchanging between clients) and so it might make sense to consolidate ways to exchange this type of data across projects.
> 
> Regarding accounts, not sure if saving account seed is enough to easily derive the key pair info but if it is, I agree with Bobbin's proposal regarding saving a specific file per account.

Yes Nullifiers implement `Serializable` and `Deserializable` but they do not implement serde `Serialize` and `Deserialize`, hence why I added this PR to miden-base: 

https://github.com/0xPolygonMiden/miden-base/pull/406

Could you send me the link to the parsing of binary files for Notes on the client side. I will try to make sure that both our implementations are compatible to keep it interoperable.

The genesis file will be toml and will contain a GenesisState that will be defined as follows: 
```rust
/// Represents the state at genesis, which will be used to derive the genesis block.
#[derive(Serialize, Deserialize, Debug)]
pub struct GenesisState {
    pub accounts: Vec<Account>,
    pub version: u64,
    pub timestamp: u64,
    pub notes: Vec<Note>,
    pub nullifiers: Vec<Nullifier>,
}
```

What do you think?

**Comment by bobbinth** [comment]
> Could you explain to me why we have our own implementations of `Serializable` and `Deserializable` for `GenesisState` and `Account`?

`GenesisState` is what we serialize into `genesis.dat` file. This data is not really human-readable. We use our own implementation because we can serialize account and other structs more efficiently this way.

> `Account` also implements Serde `Serialize` and `Deserialize` which I believe that I will need to be able to parse the genesis.toml file and turn it into a valid `GenesisState` type.

We actually don't need to parse `Account` structs from `Genesis.toml`. This file will contain human-readable descriptions of the accounts. We parse these descriptions and then based on them instantiate new accounts using functions from `miden-lib`. Then, we use these to create the `GenesisState` struct and then serialize to `genesis.dat` file.

> The genesis file will be toml and will contain a GenesisState that will be defined as follows:
> 
> ```rust
> /// Represents the state at genesis, which will be used to derive the genesis block.
> #[derive(Serialize, Deserialize, Debug)]
> pub struct GenesisState {
>     pub accounts: Vec<Account>,
>     pub version: u64,
>     pub timestamp: u64,
>     pub notes: Vec<Note>,
>     pub nullifiers: Vec<Nullifier>,
> }
> ```

This won't quite work because we can't serialize `Account` and other structs in the `toml` file. Instead, we probably want to have some other struct which gets deserialized from the `toml` file (e.g., maybe `GenesisInputs`). And then we use this file to build `GenesisState` file.

`GenesisInputs` will need to implement `Serialize`/`Deserialize` from `serde`.

**Comment by bobbinth** [comment]
One other thing, I'd probably split this into 2 PRs:

1. First to get the generic structure just for accounts working (without notes and nullifiers).
2. Then add notes and nullifiers.

**Comment by phklive** [comment]
@bobbinth, a few questions: 

**General questions**:
- For now we do not have a non-fungible faucet right? 
- To generate an [`AuthScheme`]() I need to have a `pub_key`, should the user input it or should I generate it when creating the account? If I need to generate it then I need to find a way to return the private and public keypair to the user.

**Genesis.json** (These are questions regarding the example json code that you have provided)
- Why is there an `id` as an account field?
- Why is there a `seed` as an auth field? 
- Why do we provide the storage_mode? (I haven't found a way to configure the storage mode when creating the accounts)

**Progress**

As you told me I am using the following functions from miden_lib to generate different types of accounts:

- Wallet
https://github.com/0xPolygonMiden/miden-base/blob/305142524fdedb8cc1878530609256167cfe5ef1/miden-lib/src/accounts/wallets/mod.rs#L29

- Fungible faucet
https://github.com/0xPolygonMiden/miden-base/blob/305142524fdedb8cc1878530609256167cfe5ef1/miden-lib/src/accounts/faucets/mod.rs#L29C11-L29C11

I have created these two structs to manage the inputs: 
```rust
/// Represents the struct built from genesis.toml
#[derive(Debug, Deserialize)]
pub struct GenesisInput {
    pub version: u64,
    pub timestamp: u64,
    pub accounts: Vec<AccountInput>,
}

#[derive(Debug, Deserialize)]
pub struct AccountInput {
    pub account_type: AccountType,
    pub seed: String,
    pub token_symbol: Option<String>,
    pub decimals: Option<u64>,
    pub max_supply: Option<u64>,
}
```

I am able to parse the file into these structs and I am iterating over them to be able to generate the different accounts and finally the `GenesisState`. I then write the files to disk, etc... as usual. Does this approach seem on the right track iyo?

After implementing this I will go forward to add the notes and nullifiers processing.

**Comment by bobbinth** [comment]
> For now we do not have a non-fungible faucet right?

Correct!

>  To generate an `AuthScheme` I need to have a `pub_key`, should the user input it or should I generate it when creating the account? If I need to generate it then I need to find a way to return the private and public keypair to the user.

I think for simplicity, for now we can derive both private and public keys from `seed` in the `auth` section. Then these would be output together with the account data.

In the future, we could allow specifying just public key as well - and in this case, we'd assume that users generate their key pairs themselves.

> Why is there an `id` as an account field?

Yeah - it shouldn't be there. Sorry!

> Why is there a `seed` as an auth field?

As mentioned above, this is what we can use to generate keys for a given account. We could do something similar to what we did [here](https://github.com/0xPolygonMiden/miden-base/blob/main/miden-tx/tests/common/mod.rs#L105) - but use the seed to generate `KeyPair`.

> Why do we provide the storage_mode? (I haven't found a way to configure the storage mode when creating the accounts)

Yeah - we don't have it yet - so, fine to skip for now (we will add these soon though).

> I have broken down the make_genesis() function into smaller functions encapsulating their own logic. I am able to parse the file into these structs and am iterating over the accounts field of `GenesisInput` to be able to generate the different accounts. Finally I build the `GenesisState` struct from `GenesisInput`. I then write the files to disk, etc... as usual. Does this approach seem on the right track?

Yep - all looks correct. One potential change is that I'd probably use an enum for `AccountInput`. For example, something like this:
```Rust
pub enum AccountInput {
    BasicWallet(BasicWalletInputs),
    BasicFungibleFaucet(BasicFungibleFaucetInputs),
}
```

**Comment by phklive** [comment]
> > For now we do not have a non-fungible faucet right?
> 
> Correct!
> 
> > To generate an `AuthScheme` I need to have a `pub_key`, should the user input it or should I generate it when creating the account? If I need to generate it then I need to find a way to return the private and public keypair to the user.
> 
> I think for simplicity, for now we can derive both private and public keys from `seed` in the `auth` section. Then these would be output together with the account data.
> 
> In the future, we could allow specifying just public key as well - and in this case, we'd assume that users generate their key pairs themselves.
> 
> > Why is there an `id` as an account field?
> 
> Yeah - it shouldn't be there. Sorry!
> 
> > Why is there a `seed` as an auth field?
> 
> As mentioned above, this is what we can use to generate keys for a given account. We could do something similar to what we did [here](https://github.com/0xPolygonMiden/miden-base/blob/main/miden-tx/tests/common/mod.rs#L105) - but use the seed to generate `KeyPair`.
> 
> > Why do we provide the storage_mode? (I haven't found a way to configure the storage mode when creating the accounts)
> 
> Yeah - we don't have it yet - so, fine to skip for now (we will add these soon though).
> 
> > I have broken down the make_genesis() function into smaller functions encapsulating their own logic. I am able to parse the file into these structs and am iterating over the accounts field of `GenesisInput` to be able to generate the different accounts. Finally I build the `GenesisState` struct from `GenesisInput`. I then write the files to disk, etc... as usual. Does this approach seem on the right track?
> 
> Yep - all looks correct. One potential change is that I'd probably use an enum for `AccountInput`. For example, something like this:
> 
> ```rust
> pub enum AccountInput {
>     BasicWallet(BasicWalletInputs),
>     BasicFungibleFaucet(BasicFungibleFaucetInputs),
> }
> ```

Also, I forgot: 

In the JSON code snippet you provided the `seed` as a String or Number. In code the seed is a [u8;32], hence should I request for the json code to be a list of 32 u8's for parsing or derive the seed from a String or Number? (Which I don't yet have an idea how to do).


What you provided:
```json
"seed": "0x1245",
```

What I thought it should be: 
```json 
"seed": [1,2,3,4,5,6,7,8,9...]
```

**Comment by bobbinth** [comment]
> In the JSON code snippet you provided the `seed` as a String or Number. In code the seed is a [u8;32], hence should I request for the json code to be a list of 32 u8's for parsing or derive the seed from a String or Number?

I think a hex representation (a string) would be better to have in JSON as it is more human-readable. So, you can interpret it as a hex string when converting to `[u8; 32]`.

**Comment by Dominik1999** [comment]
Is this closed by https://github.com/0xPolygonMiden/miden-node/pull/138 ?

**Comment by phklive** [comment]
> Is this closed by #138 ?

Yes once it is merged.
